lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -133,14 +133,47 @@ def safe_change_column_default(table_name, column_name, default_value)
   end
 
   def safe_make_column_nullable(table, column)
+    quoted_table_name = connection.quote_table_name(table)
+    quoted_column_name = connection.quote_column_name(column)
+
     safely_acquire_lock_for_table(table) do
-      raw_execute "ALTER TABLE #{table} ALTER COLUMN #{column} DROP NOT NULL"
+      raw_execute "ALTER TABLE #{quoted_table_name} ALTER COLUMN #{quoted_column_name} DROP NOT NULL"
+    end
+  end
+
+  # Postgres 12+ can use a valid CHECK constraint to validate that no values of a column are null, avoiding
+  # locking the table when altering a column to NOT NULL
+  #
+  # Source:
+  # https://dba.stackexchange.com/questions/267947/how-can-i-set-a-column-to-not-null-without-locking-the-table-during-a-table-scan/268128#268128
+  # (https://archive.is/X55up)
+  def safe_make_column_not_nullable(table, column)
+    if ActiveRecord::Base.connection.postgresql_version < 12_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Cannot safely make a column non-nullable before Postgres 12"
+    end
+
+    validated_table = PgHaMigrations::Table.from_table_name(table)
+    tmp_constraint_name = "tmp_not_null_constraint_#{OpenSSL::Digest::SHA256.hexdigest(column.to_s).first(7)}"
+
+    if validated_table.has_constraint?(tmp_constraint_name)
```

Comment:
Is it worth checking to see if there's a constraint on this column whether or not this method was the source of it? Or should we just ignore that since we weren't responsible for it?

I supposed we could add that check in when we add a method for converting an existing check constraint to a `NOT NULL` constraint...but both the `INVALID` and `VALID` cases technically are relevant here.

In fact, if we had a valid one already (not created by us even), we might just skip the check constraint creation. Not sure if that's worth it or not if we have a separate API for that.

I think I'm talking myself into "only this check is necessary for now, because it's the only thing that prevents us from proceeding, but something to chew on.

Reply:
Yeah I think I agree with your assessment. Would you mind opening an issue to track this new method and the enhancements described above?